### ISSUE TITLE
Help with gradient memory issue

### ISSUE THREAD

**Issue by Miguel-Fuentes** [google/jax#4367]
I have set up a model with a loss function and I am troubleshooting some code to implement basic gradient descent. The code looks like this

```
loss_and_grad = jit(value_and_grad(flow_model.compute_loss, argnums=[0,1]))
init, transition = (model.initial, model.transition)
d_matrix = model.distance_matrix
    
for i in range(n):
    loss, grad = loss_and_grad(init, transition, true_densities, d_matrix)
    init_grad, transition_grad = grad
    init = init - (lr * init_grad)
    transition = transition - (lr*transition_grad)
```

When I run this for n=1 I get a result which looks correct in about 50 seconds. When I try to run it for n > 2 the memory blows up and it crashes. For some context about the scale, init has size (2832,) and transition has size (51, 2832, 2832). The loss function on it's own (no gradient) is a jit decorated function which takes about 5 seconds on my machine with the same parameters.

Is there something wrong with how I'm implementing this which is causing the memory to blow up? I'm pretty new to jax so it might be something simple. Let me know if posting the rest of the code would help.

**Comment by jekbradbury** [comment]
Yes, it would be helpful to see the rest of the code/a runnable example (also, whether you're running this on CPU, GPU, or TPU). (A colab link is particularly convenient, if that's what you're using.)

I can't think of a good reason why N=2 should be taking more memory than N=1, or why computing the gradient + update should take 10x as long as just the forward pass.

**Comment by Miguel-Fuentes** [comment]
I've been running the code locally on a CPU. I thought the amount of time for the gradient update was suspicious.

Here is the model code. I made a bunch of static methods so that the methods operate only on objects that jit understands. I'll spin up a collab notebook and run it there as well so that I can share a link with you.  

```
from functools import partial

import jax.numpy as jnp
from jax.numpy import linalg as LA
from jax.nn import softmax
from jax.ops import index, index_update
from jax import jit, random, grad, value_and_grad
from jax.lax import scan


class flow_model:
    def __init__(self, cells, weeks, x_dim=None, rng_key=0):
        '''
        Sets several constants like number of positions to flow between, the number of weeks
        the x dimmension of the grid and the rng key. Also initialized parameters
        '''
        self.cells = cells
        self.weeks = weeks
        self.x_dim = x_dim if x_dim else jnp.sqrt(cells)
        
        self.distance_matrix = flow_model.gen_distance_matrix(self.cells, self.x_dim)
        
        self.key = key = random.PRNGKey(rng_key)
        self.initialize_params()
        
        
    def initialize_params(self, initial=None, transition=None):
        if initial is not None:
            self.initial = initial
        else:
            self.key, subkey1 = random.split(self.key)
            self.initial = random.normal(subkey1, shape=(self.cells,))
        if transition is not None:
            self.transition = transition
        else:
            self.key, subkey2 = random.split(self.key)
            self.transition = random.normal(subkey2, shape=(self.weeks - 1, self.cells, self.cells))
        
        
    def predict(self):
        '''Uses the current parameters to compute the densities and the transitions for every week'''
        return flow_model.compute_predictions(self.initial, self.transition)
    
    
    def loss(self, true_densities):
        '''Gets prediction and computes loss based on flow smoothness and matching the true densities'''
        return flow_model.compute_loss(self.initial, self.transition, true_densities, self.distance_matrix)
    
    
    @staticmethod
    @jit
    def compute_loss(initial_params, transition_params, true_densities, distance_matrix):
        pred_densities, pred_transitions = flow_model.compute_predictions(initial_params, transition_params)
        obs_loss = flow_model.observation_loss(pred_densities, true_densities)
        smooth_loss = flow_model.smoothness_loss(pred_transitions, distance_matrix)
        return obs_loss + smooth_loss
    
    
    @staticmethod
    @jit
    def compute_predictions(initial_params, transition_params):
        cells = initial_params.shape[0]
        weeks = transition_params.shape[0]
        
        pred_densities = jnp.zeros((weeks + 1, cells))
        
        pred_densities = index_update(pred_densities, index[0], softmax(initial_params, axis=0))
        carry = (pred_densities, 0)
        
        carry, pred_transitions = scan(flow_model.next_week, carry, transition_params)
        pred_densities, _ = carry
        return pred_densities, pred_transitions
    
    
    @staticmethod
    def next_week(carry, transition_params):
        densities, week = carry
        last_week = densities[week]
        
        transition_prop = softmax(transition_params, axis=1)
        transition_amount = transition_prop * last_week.reshape(-1,1)
        next_density = jnp.sum(transition_amount, axis=0)
        
        week = week + 1
        densities = index_update(densities, index[week], next_density)
        carry = (densities, week)
        return carry, transition_amount
    
    
    @staticmethod
    @partial(jit, static_argnums=(0,1))
    def gen_distance_matrix(cells, x_dim):
        '''Computes a distance matrix between every position'''
        y_dim = int(cells / x_dim)

        one_col = jnp.linspace(0, y_dim - 1, num=y_dim)
        y_val = jnp.repeat(one_col, x_dim).reshape((cells, 1))
        y_diff = y_val - y_val.T

        one_row = jnp.linspace(0, x_dim - 1, num=x_dim)
        x_val = jnp.tile(one_row, y_dim).reshape((cells, 1))
        x_diff = x_val - x_val.T

        return jnp.sqrt(jnp.square(x_diff) + jnp.square(y_diff))


    @staticmethod
    def observation_loss(pred_densities, true_densities):
        residual = true_densities - pred_densities
        return LA.norm(residual)**2


    @staticmethod
    def smoothness_loss(pred_transitions, distance_matrix):
        smooth_matrix = jnp.multiply(pred_transitions, distance_matrix)
        return jnp.sum(smooth_matrix)
```

**Comment by Miguel-Fuentes** [comment]
I just put everything in a colab instance (still using CPU) [https://colab.research.google.com/drive/1zyH9OmXv5aUNGUuKwjPpgJoiPJVXt8e7?usp=sharing](url). 

On colab the code runs much faster than on my laptop and I can look at the ram usage there. I can run it on colab for many more iterations and it doesn't seem to ever blow up, it seems to stay between 6 and 10 gb of ram usage and each iteration seems to take around 13 seconds. 

My laptop has 16 gb of ram so I still don't really understand the crashes and why it is so slow. I am on a windows machine using WSL 2 so maybe there is some resource allocation problem related to that?

**Comment by jekbradbury** [comment]
Yes, it could be an issue related to WSL. We don't have any tests or benchmarks that run on WSL, so that setting is somewhat experimental/at-your-own-risk for now.

(Anything I/we can do to debug would start with reproducing in our environment, which appears not to have this problem, but if you run into similar issues in Colab or on Linux/macOS we would definitely be able to take a look.)

**Comment by Miguel-Fuentes** [comment]
Okay, I think it has to do with WSL limiting RAM allocation. I'll close the issue because it doesn't seem like it's coming from jax.

**Comment by Miguel-Fuentes** [comment]
Actually, I think I celebrated too soon by assuming there wouldn't be any issues with colab. I've been running it on colab now (same link as before) and I will occasionally get a crash with this error message. 

RuntimeError                              Traceback (most recent call last)
<ipython-input-10-14d8a96daa0b> in <module>()
----> 1 optim_i, optim_t = time_gradient(model, true_densities, n=100)

5 frames
/usr/local/lib/python3.6/dist-packages/jax/interpreters/xla.py in _execute_compiled(compiled, handlers, *args)
    757   device, = compiled.local_devices()
    758   input_bufs = [device_put(x, device) for x in args if x is not token]
--> 759   out_bufs = compiled.execute(input_bufs)
    760   if FLAGS.jax_debug_nans: check_nans(xla_call_p, out_bufs)
    761   return [handler(out_buf) for handler, out_buf in zip(handlers, out_bufs)]

RuntimeError: Resource exhausted: Out of memory while trying to allocate 4984907648 bytes.

I seem to be able to run a cell which gets the gradient hundreds of times without an issue but if I rerun the cell more than once without resetting the runtime I get this error. Do you think this is related to the past issue or does this have something to do with rerunning the same notebook cell repeatedly?

**Comment by shoyer** [comment]
How many scanned steps are being made? Keep in mind that differentiating through scan currently requires storing all intermediate results in memory.

Note that you can tradeoff additional compute for reduced memory usage by using jax.checkpoint: https://jax.readthedocs.io/en/latest/jax.html#jax.checkpoint

**Comment by Miguel-Fuentes** [comment]
There are 51 scanned steps being made. The carry for the scan is one big array which I update at every iteration, does that mean that I am keeping track of all 51 versions of the same big array? That could definitely be part of the problem. I might be able to restructure my scan to avoid this.

Also, I didn't know about jax.checkpoint, putting that on the next_week might improve things.

**Comment by mattjj** [comment]
> The carry for the scan is one big array which I update at every iteration, does that mean that I am keeping track of all 51 versions of the same big array?

Yes, it does mean all those copies. If your access pattern is such that you can scan over the big array, rather than keeping it in the carry, that can be better, but I haven't looked at the above code to know whether that's relevant.

@apaszke just started thinking about a scan with builtin recursive checkpointing; maybe this is another potential application.

**Comment by Miguel-Fuentes** [comment]
I think I can refactor things to pull the big array out of the carry. I'll try that later today and report back on how that changes things. If that isn't enough of an improvement I'll also look at adding a checkpoint.

**Comment by Miguel-Fuentes** [comment]
I went ahead and tried everyone's suggestions in colab.

I tried to refactor the scan so that the big array isn't in the carry. I did it by adding it as another parameter to the scan function and then using partial to assign it ahead of time. I'm not sure this actually did anything because I still have to access it during the scan so I suspect that this change is surface level and the code is executing in pretty much the same way. I don't know how to remedy this though.

I tried adding a checkpoint before the scanning function but that actually seems to use more memory? It's not super exact because I am just looking at the RAM numbers colab gives at runtime for spikes. But, having done it a few times it seems like without the checkpoint the RAM spikes tend to be around 10gb but with the checkpoints it tends to spike around 11.45gb.
